File: common/src/main/java/org/bread_experts_group/breadlib/registry/BlockEntityTypeBuilder.java
package org.bread_experts_group.breadlib.registry;

import net.minecraft.client.renderer.blockentity.BlockEntityRendererProvider;
import net.minecraft.client.renderer.blockentity.BlockEntityRenderers;
import net.minecraft.world.level.block.Block;
import net.minecraft.world.level.block.entity.BlockEntity;
import net.minecraft.world.level.block.entity.BlockEntityType;

public class BlockEntityTypeBuilder<T extends BlockEntity> {
	private final BlockEntityType<T> type;

	private BlockEntityTypeBuilder(BlockEntityType<T> type) {
		this.type = type;
	}

	public static <T extends BlockEntity> BlockEntityTypeBuilder<T> create(BlockEntityType.BlockEntitySupplier<T> factory, Block... validBlocks) {
		BlockEntityType<T> type = BlockEntityType.Builder.of(factory, validBlocks).build(null);
		return new BlockEntityTypeBuilder<>(type);
	}

	public BlockEntityTypeBuilder<T> withRenderer(BlockEntityRendererProvider<T> provider) {
		BlockEntityRenderers.register(this.type, provider);
		return this;
	}

	public BlockEntityType<T> build() {
		return type;
	}
}

File: common/src/main/kotlin/org/bread_experts_group/breadlib/registry/BlockEntityTypeBuilder.kt
package org.bread_experts_group.breadlib.registry

import net.minecraft.client.renderer.blockentity.BlockEntityRendererProvider
import net.minecraft.client.renderer.blockentity.BlockEntityRenderers
import net.minecraft.world.level.block.Block
import net.minecraft.world.level.block.entity.BlockEntity
import net.minecraft.world.level.block.entity.BlockEntityType

class BlockEntityTypeBuilder<T : BlockEntity> private constructor(private val type: BlockEntityType<T>) {
	companion object {
		@JvmStatic
		fun <T : BlockEntity> create(
			factory: BlockEntityType.BlockEntitySupplier<T>,
			vararg validBlocks: Block
		): BlockEntityTypeBuilder<T> {
			val type = BlockEntityType.Builder.of<T>(factory, *validBlocks).build(null)
			return BlockEntityTypeBuilder<T>(type)
		}
	}

	fun withRenderer(provider: BlockEntityRendererProvider<T>): BlockEntityTypeBuilder<T> {
		BlockEntityRenderers.register(this.type, provider)
		return this
	}

	fun build(): BlockEntityType<T> = type
}